File: experimental/flutter_plugin/hidden_volume/bench/ffi_overhead_bench.dart
// ignore_for_file: avoid_print
// Micro-benchmark for the dart:ffi overhead of the hidden_volume sync
// bindings. Compares against the same operations driven from Python
// (`bench/ffi_overhead_bench.py`) — the workload is identical down to
// the chunk count.
//
// Run from repo root:
//   dart run experimental/flutter_plugin/hidden_volume/bench/ffi_overhead_bench.dart
// (Use a host with the cdylib already built at target/release.)
//
// What this measures:
//   1. SpaceHandle.create — dominated by Argon2 KDF (~30 ms light)
//   2. commit  — 1 KV put per call, includes one chunk write + fsync
//   3. get     — read one KV value
//   4. headerInfo — password-less header read (LOCK_SH)
//
// We deliberately use small payloads so per-call FFI overhead, not
// payload serialization, dominates the time budget.

import 'dart:io';
import 'dart:typed_data';

import 'package:hidden_volume/src/bindings.dart';

const int _opsPerSample = 200;
const int _samples = 5;

void main() {
  final dllPath = _resolveDylibPath();
  overrideDylib(DynamicLibrary.open(dllPath));
  print('cdylib: $dllPath');
  print('uniffi contract version: ${contractVersion()}');
  print('');

  final tmp = Directory.systemTemp.createTempSync('hv_bench_');
  try {
    _benchmarkCreate(tmp);
    _benchmarkGetCommit(tmp);
    _benchmarkHeaderInfo(tmp);
  } finally {
    tmp.deleteSync(recursive: true);
  }
}

void _benchmarkCreate(Directory tmp) {
  // 5 cold creates, each with ArgonPreset.light. Don't measure single
  // calls — Argon2 dominates and that's a Rust cost, not an FFI cost.
  // But report it anyway so the user has the absolute number.
  final samples = <int>[];
  for (var i = 0; i < _samples; i++) {
    final path = '${tmp.path}/create_$i.bin';
    final sw = Stopwatch()..start();
    final s = SpaceHandleBindings.create(
      path: path,
      password: Uint8List.fromList('bench'.codeUnits),
      argon: ArgonPreset.light,
    );
    sw.stop();
    s.close();
    samples.add(sw.elapsedMicroseconds);
  }
  _report('create (Argon2 light + 1 init commit)', samples, 1);
}

void _benchmarkGetCommit(Directory tmp) {
  final path = '${tmp.path}/getcommit.bin';
  final space = SpaceHandleBindings.create(
    path: path,
    password: Uint8List.fromList('bench'.codeUnits),
    argon: ArgonPreset.light,
  );

  // Pre-populate keys so the get() loop has something to read.
  final keys = <Uint8List>[
    for (var i = 0; i < _opsPerSample; i++)
      Uint8List.fromList('k$i'.codeUnits),
  ];
  space.commit([
    for (var i = 0; i < _opsPerSample; i++)
      HvWriteOpPut(
        namespace: 1,
        key: keys[i],
        value: Uint8List.fromList('v$i'.codeUnits),
      ),
  ]);

  // get() — 200 reads per sample
  final getSamples = <int>[];
  for (var s = 0; s < _samples; s++) {
    final sw = Stopwatch()..start();
    for (var i = 0; i < _opsPerSample; i++) {
      space.get(1, keys[i]);
    }
    sw.stop();
    getSamples.add(sw.elapsedMicroseconds);
  }
  _report('get  (200 reads/sample)', getSamples, _opsPerSample);

  // commit() — 200 single-put commits per sample. Each includes a
  // chunk write + fsync, so this is upper-bounded by disk speed.
  final commitSamples = <int>[];
  for (var s = 0; s < _samples; s++) {
    final sw = Stopwatch()..start();
    for (var i = 0; i < _opsPerSample; i++) {
      space.commit([
        HvWriteOpPut(
          namespace: 2,
          key: Uint8List.fromList('c${s}_$i'.codeUnits),
          value: Uint8List.fromList('v'.codeUnits),
        ),
      ]);
    }
    sw.stop();
    commitSamples.add(sw.elapsedMicroseconds);
  }
  _report('commit (1 KV + fsync per call)', commitSamples, _opsPerSample);

  space.close();
}

void _benchmarkHeaderInfo(Directory tmp) {
  final path = '${tmp.path}/header.bin';
  final s = SpaceHandleBindings.create(
    path: path,
    password: Uint8List.fromList('bench'.codeUnits),
    argon: ArgonPreset.light,
  );
  s.close();

  final samples = <int>[];
  for (var sample = 0; sample < _samples; sample++) {
    final sw = Stopwatch()..start();
    for (var i = 0; i < _opsPerSample; i++) {
      headerInfo(path);
    }
    sw.stop();
    samples.add(sw.elapsedMicroseconds);
  }
  _report('headerInfo (200 reads/sample)', samples, _opsPerSample);
}

void _report(String label, List<int> sampleMicros, int opsPerSample) {
  final perOpUs = sampleMicros.map((m) => m / opsPerSample).toList()..sort();
  final min = perOpUs.first;
  final p50 = perOpUs[perOpUs.length ~/ 2];
  final max = perOpUs.last;
  print('${label.padRight(36)}'
      '  per-op: min=${min.toStringAsFixed(2)}µs  '
      'p50=${p50.toStringAsFixed(2)}µs  '
      'max=${max.toStringAsFixed(2)}µs  '
      '($_samples samples × $opsPerSample ops)');
}

String _resolveDylibPath() {
  final root = Directory.current.path;
  // From repo root OR from the plugin dir — try both depths.
  final candidates = <String>[
    '$root/target/release/hidden_volume_ffi.dll',
    '$root/target/debug/hidden_volume_ffi.dll',
    '$root/../../../target/release/hidden_volume_ffi.dll',
    '$root/../../../target/debug/hidden_volume_ffi.dll',
  ];
  for (final p in candidates) {
    if (File(p).existsSync()) return p;
  }
  throw StateError('cdylib not found, searched: ${candidates.join(", ")}');
}

File: experimental/flutter_plugin/hidden_volume/bench/ffi_overhead_bench.py
"""Python ctypes counterpart of `ffi_overhead_bench.dart`.

Runs the same workload (create / get / commit / header_info) so that
the dart:ffi numbers can be compared against a reference ctypes-based
binding driving the same uniffi 0.31 cdylib.

Run from the repo root:
    python experimental/flutter_plugin/hidden_volume/bench/ffi_overhead_bench.py
"""

from __future__ import annotations

import os
import shutil
import sys
import tempfile
import time
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[4]
BINDINGS_PY = REPO_ROOT / "bindings" / "python"

# Ensure the cdylib is co-located next to hidden_volume_ffi.py so the
# stock loader (`uniffi_lib_init`) finds it on Windows.
DLL_NAME = "hidden_volume_ffi.dll"
SRC_DLL = REPO_ROOT / "target" / "release" / DLL_NAME
DST_DLL = BINDINGS_PY / DLL_NAME
if not DST_DLL.exists() or DST_DLL.stat().st_mtime < SRC_DLL.stat().st_mtime:
    shutil.copy2(SRC_DLL, DST_DLL)

sys.path.insert(0, str(BINDINGS_PY))
import hidden_volume_ffi as hv  # noqa: E402

OPS_PER_SAMPLE = 200
SAMPLES = 5


def report(label: str, sample_secs: list[float], ops_per_sample: int) -> None:
    per_op_us = sorted(s / ops_per_sample * 1e6 for s in sample_secs)
    print(
        f"{label:<36}  per-op: "
        f"min={per_op_us[0]:.2f}us  "
        f"p50={per_op_us[len(per_op_us)//2]:.2f}us  "
        f"max={per_op_us[-1]:.2f}us  "
        f"({SAMPLES} samples x {ops_per_sample} ops)"
    )


def bench_create(tmp: Path) -> None:
    samples: list[float] = []
    for i in range(SAMPLES):
        path = tmp / f"create_{i}.bin"
        t0 = time.perf_counter()
        s = hv.SpaceHandle.create(
            path=str(path),
            password=b"bench",
            argon=hv.ArgonPreset.LIGHT,
            initial_garbage_chunks=0,
            superblock_replicas=3,
        )
        samples.append(time.perf_counter() - t0)
        del s  # __del__ frees the handle
    report("create (Argon2 light + 1 init commit)", samples, 1)


def bench_get_commit(tmp: Path) -> None:
    path = tmp / "getcommit.bin"
    space = hv.SpaceHandle.create(
        path=str(path),
        password=b"bench",
        argon=hv.ArgonPreset.LIGHT,
        initial_garbage_chunks=0,
        superblock_replicas=3,
    )
    keys = [f"k{i}".encode() for i in range(OPS_PER_SAMPLE)]
    space.commit([
        hv.WriteOp.PUT(namespace=1, key=k, value=f"v{i}".encode())
        for i, k in enumerate(keys)
    ])

    get_samples: list[float] = []
    for _ in range(SAMPLES):
        t0 = time.perf_counter()
        for k in keys:
            space.get(namespace=1, key=k)
        get_samples.append(time.perf_counter() - t0)
    report("get  (200 reads/sample)", get_samples, OPS_PER_SAMPLE)

    commit_samples: list[float] = []
    for s in range(SAMPLES):
        t0 = time.perf_counter()
        for i in range(OPS_PER_SAMPLE):
            space.commit([
                hv.WriteOp.PUT(
                    namespace=2,
                    key=f"c{s}_{i}".encode(),
                    value=b"v",
                )
            ])
        commit_samples.append(time.perf_counter() - t0)
    report("commit (1 KV + fsync per call)", commit_samples, OPS_PER_SAMPLE)
    del space


def bench_header_info(tmp: Path) -> None:
    path = tmp / "header.bin"
    s = hv.SpaceHandle.create(
        path=str(path),
        password=b"bench",
        argon=hv.ArgonPreset.LIGHT,
        initial_garbage_chunks=0,
        superblock_replicas=3,
    )
    del s
    samples: list[float] = []
    for _ in range(SAMPLES):
        t0 = time.perf_counter()
        for _ in range(OPS_PER_SAMPLE):
            hv.header_info(path=str(path))
        samples.append(time.perf_counter() - t0)
    report("header_info (200 reads/sample)", samples, OPS_PER_SAMPLE)


def main() -> None:
    print(f"cdylib: {SRC_DLL}")
    print()
    with tempfile.TemporaryDirectory(prefix="hv_bench_") as d:
        tmp = Path(d)
        bench_create(tmp)
        bench_get_commit(tmp)
        bench_header_info(tmp)


if __name__ == "__main__":
    os.environ.setdefault("PYTHONIOENCODING", "utf-8")
    main()
